Mathlib/Tactic/Linter/DeprecatedModule.lean

```diff
@@ -0,0 +1,137 @@
+* a module `Name` that has been deprecated and
+* an array of `Name`s of modules that should be imported instead
+
+to the environment.
+-/
+initialize deprecatedModuleExt :
+    SimplePersistentEnvExtension (Name × Array Name) (Std.HashSet (Name × Array Name)) ←
+  registerSimplePersistentEnvExtension {
+    addImportedFn := (·.foldl Std.HashSet.insertMany {})
+    addEntryFn := .insert
+  }
+
+/--
+`addModuleDeprecation` adds to the `deprecatedModuleExt` extension the pair consisting of the
+current module name and the array of its direct imports.
+
+It ignores the `Init` import, since this is a special module that is expected to be imported
+by all files.
+-/
+def addModuleDeprecation {m : Type → Type} [Monad m] [MonadEnv m] [MonadQuotation m] : m Unit := do
+  modifyEnv (deprecatedModuleExt.addEntry ·
+    (← getMainModule, (← getEnv).imports.filterMap fun i =>
+      if i.module == `Init then none else i.module))
+
+/--
+`deprecated_module since yyyy-mm-dd` deprecates the current module `A` in favour of
+its direct imports.
+This means that any file that directly imports `A` will get a notification on the `import A` line
+suggesting to instead import the *direct imports* of `A`.
+-/
+elab (name := deprecated_modules)
+    "deprecated_module " &"since " yyyy:num "-" mm:num "-" dd:num : command => do
+  if yyyy.getNat < 2025 then
+    throwErrorAt yyyy "The year should be at least 2025!"
+  if mm.getNat == 0 || 12 < mm.getNat || mm.raw.getSubstring?.get!.toString.trim.length != 2 then
+    throwErrorAt mm "The month should be of the form 01, 02, ..., 12!"
+  if dd.getNat == 0 || 31 < dd.getNat || dd.raw.getSubstring?.get!.toString.trim.length != 2 then
+    throwErrorAt dd "The day should be of the form 01, 02, ..., 31!"
+  addModuleDeprecation
+  -- disable the linter, so that it does not complain in the file with the deprecation
+  elabCommand (← `(set_option linter.deprecated.module false))
+
+/-- A utility function to show the pairings `(deprecatedModule, #[preferredModules])`. -/
+elab "#show_deprecated_modules" : command => do
+  let directImports := deprecatedModuleExt.getState (← getEnv)
+  logInfo <| "\n".intercalate <|
+    directImports.fold (init := ["Deprecated modules\n"]) fun nms (i, deps) =>
+      nms ++ [s!"'{i}' deprecates to\n{deps}\n"]
+
+namespace DeprecatedModule
+
+@[inherit_doc Mathlib.Linter.linter.deprecated.module]
+def deprecated.moduleLinter : Linter where run := withSetOptionIn fun stx ↦ do
+  let deprecations := deprecatedModuleExt.getState (← getEnv)
+  if deprecations.isEmpty then
+    return
+  unless Linter.getLinterValue linter.deprecated.module (← getOptions) do
+    return
+  if (← get).messages.hasErrors then
+    return
+  let mainModule ← getMainModule
+  -- `Mathlib.lean` and `Mathlib/Tactic.Lean` are allowed to import deprecated files,
+  -- since they (are expected to) import all modules in certain directories.
+  if #[`Mathlib, `Mathlib.Tactic].contains mainModule then return
```

Comment:
This is going to be annoying for downstream libraries which also provide deprecated modules.

Reply:
Is the fact that the linter can be disabled via `set_option` not a viable solution to this?

Reply:
With the `IO.Ref` to detect the "first command" of a file, it may be simple to get the linter to never flag import-only files: if the `IO.Ref` is `false` and the currently parsed command is terminal, then the linter can be instructed to do nothing.  Otherwise, the file contains a non-import command, the linter latches onto that one and does its check.

Would this be a better alternative to singling out `Mathlib.lean` and `Mathlib/Tactic.lean`?

Reply:
I implemented this: if the current command is "terminal" and the `IO.Ref` is still `false`, then the linter does nothing.  Since, upon finding any non-import, non-terminal command, the first action of the linter is to set the `IO.Ref` to `true`, this means that import-only files are automatically ignored by the linter, thus not triggering a deprecation.

This is useful for `Mathlib.lean` and `Tactic.lean` that would still contain deprecated imports, but should be "quiet".